### ISSUE TITLE
Hibernate Query Cache using excessive Heap

### ISSUE THREAD

**Issue by robinroos** [ikasanEIP/ikasan#948]
Prior to Hibernate 5.2.12, CriteriaQuery always used SQL literals for numeric parameters,  This makes the Hibernate Query Cache record every query which includes a numeric value as being one instance of a discrete query (the value is not parameterised).

Since Hibernate 5.2.12 this feature is parameterisable in code, and since 5.2.13 it is parameterisable by config.  To use placeholders for Numeric parameters to CriteriaQuery instances set:

`hibernate.criteria.literal_handling_mode=BIND`

Without this setting, the default value INLINE applies.  

This Ikasan issue is raised as a result of a JProfiler Heap Dump analysis which showed > 1 million Integer instances being reachable from Hibernate's QueryPlanCache.

![QueryPlanCache](https://user-images.githubusercontent.com/13133023/124488296-968c8300-dda7-11eb-812c-b1f332f1e43a.PNG)

**Comment by robinroos** [comment]
Below are some internet resources related to this issue which may be of interest.

https://vladmihalcea.com/how-does-hibernate-handle-jpa-criteria-api-literals/

https://hibernate.atlassian.net/browse/HHH-9576

https://hibernate.atlassian.net/browse/HHH-4698

**Comment by robinroos** [comment]
We analyzed heap of a running Ikasan module because we were trying to get them (modules plural) to run with memory footprints as small as possible.  This particular module was failing with memory issues if it had less that 1Gb of heap allocated to it, which was felt to be excessive.

Initial heap analysis showed > 1 million Integer issues retained on heap after GC, which warranted further investigation.

Subsequent analysis of incoming references to a handful of these Integers showed that they were referenced from the Hibernate QueryPlanCache, specifically from SQL SELECT statements of the form:

`select ... from ... where Expiry < 1625485380064`

e.g.
![MessageFilter](https://user-images.githubusercontent.com/13133023/124596219-ecb9fe80-de59-11eb-9228-38850d1ff735.PNG)

Clearly the Integer literal, representing a specific time, had no place being in the QueryCache.

A review of Ikasan source did not yield any problematic query invocations - indeed everything looked to be efficiently written.  Subsequent internet searches around the topic yielded the links above which led to the solution.

We have re-run the same module under load.  The number of Integer instance retained on heap after GC has dropped to just 10k.  Further indication that the fix is working can be seen in the SQL which is cached, and which has ? placeholders instead of Integer literals:

![image](https://user-images.githubusercontent.com/13133023/124596672-75d13580-de5a-11eb-8b15-a01e2d559d40.png)

The number of QueryLoader instances is reduced from approx 3k to just 16, showing that the identified memory leak has been resolved.

This issue and associated fix is not about reducing the number of Integers on the heap.  It's about optimising Hibernate so that parameterised CriteriaQuery invocations are emitted as prepared statements which can be efficiently reused and therefore warrent their place in the query cache, serving to reduce heap usage (just 16 query plans on heap) and improve performance (through increased hits on the query cache, obviating the need to compile those queries again).  Since our focus was heap utilisation, which stands resolved, we have not performed any performance-focused profiling.
